File: src/main/java/february_2025/LRUCache.java
package february_2025;

import java.util.HashMap;
import java.util.Map;

public class LRUCache {

    public static void main(String[] args) {
        LRUCache lruCache = new LRUCache(2);
        lruCache.put(1, 1);
        lruCache.put(2, 2);
        int result = lruCache.get(1);
        System.out.println(result);
        lruCache.put(3, 3);
        int result1 = lruCache.get(2);
        System.out.println();
    }

    Map<Integer, Node> cache;
    Node head;
    Node tail;
    int capacity;

    public LRUCache(int capacity) {
        this.cache = new HashMap<>();
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        this.head.prev = this.tail;
        this.tail.next = this.head;
        this.capacity = capacity;
    }

    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        Node node = cache.get(key);
        // move to front
        removeNode(node);
        addNode(node);
        return node.val;
    }

    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            Node currNode = cache.get(key);
            currNode.val = value;
            removeNode(currNode);
            addNode(currNode);
            return;
        }
        Node newNode = new Node(key, value);
        addNode(newNode);
        cache.put(key, newNode);
        if (cache.size() > capacity) {
            Node lastNode = this.tail.next;
            removeNode(lastNode);
            cache.remove(lastNode.key);
        }
    }

    private void addNode(Node node) {
        // node is empty
        if (head.prev == tail) {
            node.next = this.head;
            node.prev = this.tail;
            this.head.prev = node;
            this.tail.next = node;
        } else {
            Node prevNode = this.head.prev;
            this.head.prev = node;
            prevNode.next = node;
            node.next = this.head;
            node.prev = prevNode;
        }
    }

    private void removeNode(Node node) {
        // one node
        if (node.prev == this.tail && node.next == this.head) {
            this.head.prev = this.tail;
            this.tail.next = this.head;
        } else {
            Node prev = node.prev;
            Node next = node.next;
            next.prev = prev;
            prev.next = next;
        }
    }

    static class Node {
        int val;
        int key;
        Node prev;
        Node next;

        public Node(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }

}

File: python/lru_cache.py
class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.cache = {}
        self.capacity = capacity
        self.tail = Node(0, -1)
        self.head = Node(0, -2)
        self.tail.next = self.head
        self.head.prev = self.tail
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if not key in self.cache:
            return -1
    
        node = self.cache[key]
        # remove node
        self.removeNode(node)
        # add to front
        self.addNode(node)
        return node.value
        

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.cache:
            node = self.cache[key]
            self.removeNode(node)
        newNode = Node(value, key)
        self.addNode(newNode)
        self.cache[key] = newNode

        if len(self.cache) > self.capacity:
            tailNode = self.tail.next
            self.removeNode(tailNode)
            del self.cache[tailNode.key]

    def removeNode(self, node):
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode

    def addNode(self, newNode):
        if self.tail.next == self.head:
            # list is empty
            self.tail.next = newNode
            self.head.prev = newNode
            newNode.next = self.head
            newNode.prev = self.tail
        else:
            prevNode = self.head.prev
            prevNode.next = newNode
            self.head.prev = newNode
            newNode.next = self.head
            newNode.prev = prevNode

class Node:
    def __init__(self, value, key, prev = None, next = None):
        self.value = value
        self.key = key
        self.prev = prev
        self.next = next


solution = LRUCache(2)
solution.put(1, 1)
solution.put(2, 2)
print(solution.get(1))
solution.put(3, 3)
print(solution.get(2))
solution.put(4, 4)
print(solution.get(1))
print(solution.get(3))
print(solution.get(4))